### PR TITLE
Added script to generate a basic RevocationSet from TestNet or MainNet

### PR DESCRIPTION
**Pull request by vijs**
#### Problem
* The algorithm specified in section "6.2.4.1. Conceptual algorithm for revocation set construction" should be built and usable against TestNet schema or MainNet schema. Reference: Issue #26432

#### Change overview
* Added script to generate a basic RevocationSet from TestNet or MainNet
  * how to use it? - python ./credentials/generate-revocation-set.py --help

### CODE REVIEW

File: credentials/generate-revocation-set.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        try:
+                if isinstance(sb, str):
+                        # If there's any unicode here, an exception will be thrown and the function will return false
+                        sb_bytes = bytes(sb, 'ascii')
+                elif isinstance(sb, bytes):
+                        sb_bytes = sb
+                else:
+                        raise ValueError("Argument must be string or bytes")
+                return base64.b64encode(base64.b64decode(sb_bytes)) == sb_bytes
+        except Exception:
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:54]: I think ValueError would cover all of it

File: credentials/generate-revocation-set.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                return False
```
[the remark this answers is not in the document]
Comment (vijs) [reply credentials/generate-revocation-set.py]: Resolved this in the latest commit. Please review and share your comments.

File: credentials/generate-revocation-set.py (hunk 1 above)
Comment (vijs) [reply credentials/generate-revocation-set.py]: Addressed in the latest commit.

File: credentials/generate-revocation-set.py [hunk 3]
```diff
@@ -81,41 +68,109 @@ def main(use_main_net_dcld, use_test_net_dcld, use_main_net_http, use_test_net_h
...
-        revoked_cert_list = json.loads(cmdpipe.stdout.read())["revokedCertificates"]
+        revocation_point_list = json.loads(cmdpipe.stdout.read())["PkiRevocationDistributionPoint"]
 
     revocation_set = []
 
-    for revoked_cert in revoked_cert_list:
-        serialnumber_list = []
+    for revocation_point in revocation_point_list:
+        # 1. Validate Revocation Type
+        if revocation_point["revocationType"] != 1:
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:86]: Suggest unnesting (continue if revocationType == 1) and use enums for 1

File: credentials/generate-revocation-set.py [hunk 4]
```diff
@@ -81,41 +68,109 @@ def main(use_main_net_dcld, use_test_net_dcld, use_main_net_http, use_test_net_h
...
+
+            # 2. Parse the certificate
+            crl_signer_certificate = x509.load_pem_x509_certificate(revocation_point["crlSignerCertificate"])
+
+            vid = revocation_point["vid"]
+            pid = revocation_point["pid"]
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:92]: The VID or PID can be embedded in the subject as well. We need to factor the extraction into methods.

File: credentials/generate-revocation-set.py [hunk 5]
```diff
@@ -81,41 +68,109 @@ def main(use_main_net_dcld, use_test_net_dcld, use_main_net_http, use_test_net_h
...
+            is_paa = revocation_point["isPAA"]
+
+            # 3. && 4. Validate VID/PID
+            crl_vid_list = crl_signer_certificate.subject.get_attributes_for_oid(OID_VENDOR_ID)
+            crl_pid_list = crl_signer_certificate.subject.get_attributes_for_oid(OID_PRODUCT_ID)
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:97]: The VID or PID can be embedded in the subject as well. We need to factor the extraction into methods.

File: credentials/generate-revocation-set.py (hunk 3 above)
Comment (vijs) [reply credentials/generate-revocation-set.py]: Agreed. Fixed in commit - 1bf0a64b5105df13e56b58fa5ea457d59b3dc6b7.

File: credentials/generate-revocation-set.py (hunk 5 above)
Comment (vijs) [reply credentials/generate-revocation-set.py]: Agreed. Fixed in commit - 1bf0a64b5105df13e56b58fa5ea457d59b3dc6b7.

File: credentials/generate-revocation-set.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        # 2. Parse the certificate
+        crl_signer_certificate = x509.load_pem_x509_certificate(revocation_point["crlSignerCertificate"])
+
+        vid = revocation_point["vid"]
+        pid = revocation_point["pid"]
+        is_paa = revocation_point["isPAA"]
+
+        # 3. && 4. Validate VID/PID
+        crl_vid = extract_single_integer_attribute(crl_signer_certificate.subject, OID_VENDOR_ID)
+        crl_pid = extract_single_integer_attribute(crl_signer_certificate.subject, OID_PRODUCT_ID)
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:114]: Need to support alternate representation of VID/PID (see spec "6.2.2.2. Encoding of Vendor ID and Product ID in subject and issuer fields"). Please add a TODO. The fix itself can be done in a follow-up.

File: credentials/generate-revocation-set.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        if is_paa == True:
+            if crl_vid != None:
+                if vid != crl_vid:
+                    continue
+        else:
+            if vid == None or vid != crl_vid:
+                continue
+            if crl_pid != None:
+                if pid != crl_pid:
+                    continue
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:125]: Need to log all situations where a continue is called --> these should not happen. Can be done in a follow-up. Please add a TODO. The fix itself can be done in a follow-up.

File: credentials/generate-revocation-set.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+        # 5.
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:127]:
```suggestion
        # 5. Validate the certification path containing CRLSignerCertificate.
```

File: credentials/generate-revocation-set.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        revocation_point_list = requests.get(f"{rest_node_url}/dcl/pki/revocation-points").json()["PkiRevocationDistributionPoint"]
+    else:
+        cmdlist = ['config', 'output', 'json']
+        subprocess.Popen([dcld] + cmdlist)
+
+        cmdlist = ['query', 'pki', 'all-revocation-points']
+
+        cmdpipe = subprocess.Popen(use_dcld(dcld, production, cmdlist), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        revocation_point_list = json.loads(cmdpipe.stdout.read())["PkiRevocationDistributionPoint"]
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:96]: Extract this to a helper function. Can be done as a follow-up.

File: credentials/generate-revocation-set.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        paa_certificate = None
+
+        if use_rest:
+            response = requests.get(
+                f"{rest_node_url}/dcl/pki/certificates/{crl_signer_issuer_name}/{crl_signer_authority_key_id}").json()["approvedCertificates"]["certs"][0]
+            paa_certificate = response["pemCert"]
+        else:
+            cmdlist = ['query', 'pki', 'x509-cert', '-u', crl_signer_issuer_name, '-k', crl_signer_authority_key_id]
+            cmdpipe = subprocess.Popen(use_dcld(dcld, production, cmdlist), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            paa_certificate = json.loads(cmdpipe.stdout.read())["approvedCertificates"]["certs"][0]["pemCert"]
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:142]: Extract this operation to a helper function.  Can be done as a follow-up.

File: credentials/generate-revocation-set.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+        if use_rest:
+            response = requests.get(
+                f"{rest_node_url}/dcl/pki/revocation-points/{issuer_subject_key_id}").json()["pkiRevocationDistributionPointsByIssuerSubjectKeyID"]
+            same_issuer_points = response["points"]
+        else:
+            cmdlist = ['query', 'pki', 'revocation-points', '--issuer-subject-key-id', issuer_subject_key_id]
+            cmdpipe = subprocess.Popen(use_dcld(dcld, production, cmdlist), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            same_issuer_points = json.loads(cmdpipe.stdout.read())[
+                "pkiRevocationDistributionPointsByIssuerSubjectKeyID"]["points"]
```
Comment (tcarmelveilleux) [inline credentials/generate-revocation-set.py:177]: Extract to a helper function. Can be done in a follow-up.

**Review by tcarmelveilleux — APPROVED** [review]

File: credentials/generate-revocation-set.py (hunk 6 above)
Comment (vijs) [reply credentials/generate-revocation-set.py:116]: Added comments in the src file to capture the follow-up work.

File: credentials/generate-revocation-set.py (hunk 8 above)
Comment (vijs) [reply credentials/generate-revocation-set.py]: Addressed in the latest commit.

File: credentials/generate-revocation-set.py (hunk 10 above)
Comment (vijs) [reply credentials/generate-revocation-set.py:146]: Added comments in the src file to capture the follow-up work.

File: credentials/generate-revocation-set.py (hunk 9 above)
Comment (vijs) [reply credentials/generate-revocation-set.py:97]: Added comments in the src file to capture the follow-up work.

File: credentials/generate-revocation-set.py (hunk 11 above)
Comment (vijs) [reply credentials/generate-revocation-set.py:182]: Added comments in the src file to capture the follow-up work.

File: credentials/generate-revocation-set.py (hunk 7 above)
Comment (vijs) [reply credentials/generate-revocation-set.py]: Added comments in the src file to capture the follow-up work.

**Review by hawk248 — APPROVED** [review]

**Pull request merged**